internal/provider/atlas_schema_resource.go

```diff
@@ -0,0 +1,347 @@
+	if err != nil {
+		diags.AddError("Inspect Error", fmt.Sprintf("Unable to inspect realm, got error: %s", err))
+		return
+	}
+	hcl, err := cli.MarshalSpec(realm)
+	if err != nil {
+		diags.AddError("Marshal Error", fmt.Sprintf("Unable to marshal, got error: %s", err))
+		return
+	}
+
+	data.ID = types.String{Value: url}
+	data.URL = types.String{Value: url}
+	data.HCL = types.String{Value: string(hcl)}
+
+	// Save updated data into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+func (r *AtlasSchemaResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	var data *AtlasSchemaResourceModel
+
+	// Read Terraform plan data into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	resp.Diagnostics.Append(r.applySchema(ctx, data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Save updated data into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+func (r *AtlasSchemaResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	var data *AtlasSchemaResourceModel
+
+	// Read Terraform prior state data into the model
+	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Delete the resource by setting
+	// the HCL to an empty string
+	data.HCL = types.String{Null: true}
+	resp.Diagnostics.Append(r.applySchema(ctx, data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
+
+func (r AtlasSchemaResource) ValidateConfig(ctx context.Context, req resource.ValidateConfigRequest, resp *resource.ValidateConfigResponse) {
+	var data AtlasSchemaResourceModel
+
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	if data.DevURL.Value == "" {
+		resp.Diagnostics.AddAttributeWarning(
+			path.Root("dev_db_url"),
+			"dev_db_url is unset",
+			"It is highly recommended that you use 'dev_db_url' to specify a dev database.\n"+
+				"to learn more about it, visit: https://atlasgo.io/dev-database")
+	}
+}
+
+func (r *AtlasSchemaResource) ModifyPlan(ctx context.Context, req resource.ModifyPlanRequest, resp *resource.ModifyPlanResponse) {
+	var state *AtlasSchemaResourceModel
+
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	if state != nil && state.HCL.Value != "" {
+		// This isn't a new resource, so we don't need to do anything
+		return
+	}
+
+	var plan *AtlasSchemaResourceModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	resp.Diagnostics.Append(r.firstRunCheck(ctx, plan)...)
+}
+
+func (r *AtlasSchemaResource) applySchema(ctx context.Context, data *AtlasSchemaResourceModel) (diags diag.Diagnostics) {
+	createDesired := func(ctx context.Context, cli *sqlclient.Client) (desired *schema.Realm, err error) {
```

Comment:
why is this passed as an arg to `atlasChanges`? can't it just be part of that method?

Reply:
I reused the function for other places of the provider